### PR TITLE
add abac tenant logic validation

### PR DESCRIPTION
**Pull request by dmitrytulba**
# Changes

> Describe your changes in a short and concise list.

# Related issue

> Reference related issues, and use one of the [closing keywords, e.g. closes or fixes](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) to link the corresponding issue, if any.

# Additional information 

> Provide additional information for this change, if needed.

# Pre-Merge checklist

- [ ] New code is tested
- [ ] Present and new tests pass
- [ ] Documentation is updated
- [ ] The build is working without errors
- [ ] No new Sonar issues introduced
- [ ] Changelog is updated
- [ ] Code has been reviewed

### CODE REVIEW

File: application/src/main/java/org/ehrbase/application/abac/CustomMethodSecurityExpressionRoot.java [hunk 1]
```diff
@@ -225,35 +231,77 @@ private boolean checkAbac(String type, String subject, Object payload, String co
...
+    /**
+     * Extracts the requested claim from the token's claims.
+     * @param token Token
+     * @param requestedAbacClaim The claim to be retrieved
+     * @return The value of the requested claim
+     */
+    private String getRequestedClaim(AbstractAuthenticationToken token, String requestedAbacClaim) {
+        String claim = EMPTY;
+        Object principal = token.getCredentials();
+        if (principal instanceof Jwt) {
```
Comment (micha1970) [inline application/src/main/java/org/ehrbase/application/abac/CustomMethodSecurityExpressionRoot.java:258]: You can use in java 17 xyz instanceof ABC obj

File: application/src/main/java/org/ehrbase/application/abac/CustomMethodSecurityExpressionRoot.java [hunk 2]
```diff
@@ -225,35 +231,77 @@ private boolean checkAbac(String type, String subject, Object payload, String co
...
+            Map<String, Object> claims = ((Jwt) principal).getClaims();
+            if (claims != null && claims.containsKey(requestedAbacClaim)) {
+                claim = claims.get(requestedAbacClaim).toString();
+            }
+        } else if (principal instanceof DecodedJWT) {
```
Comment (micha1970) [inline application/src/main/java/org/ehrbase/application/abac/CustomMethodSecurityExpressionRoot.java:263]: Same here

File: application/src/main/java/org/ehrbase/application/abac/CustomMethodSecurityExpressionRoot.java [hunk 3]
```diff
@@ -225,35 +231,77 @@ private boolean checkAbac(String type, String subject, Object payload, String co
...
+     * @param token Token
+     * @param requestedAbacClaim The claim to be retrieved
+     * @return The value of the requested claim
+     */
+    private String getRequestedClaim(AbstractAuthenticationToken token, String requestedAbacClaim) {
+        String claim = EMPTY;
+        Object principal = token.getCredentials();
+        if (principal instanceof Jwt jwt) {
+            Map<String, Object> claims = jwt.getClaims();
+            if (claims != null && claims.containsKey(requestedAbacClaim)) {
```
Comment (micha1970) [inline application/src/main/java/org/ehrbase/application/abac/CustomMethodSecurityExpressionRoot.java:260]: Lines 260 and 265 are the same, move them into a Function<,>

File: application/src/main/java/org/ehrbase/application/abac/CustomMethodSecurityExpressionRoot.java [hunk 4]
```diff
@@ -307,39 +355,39 @@ private void templateHandling(
...
-                            if (body instanceof Map) {
-                                Object error = ((Map<?, ?>) body).get("error");
-                                if (error != null && ((String) error).contains("delet")) {
-                                    // composition was deleted, so nothing to check here, skip
-                                    break;
-                                }
+                    if (!(payload instanceof ResponseEntity<?>)) {
+                        throw new InternalServerException("ABAC: unexpected empty response body");
+                    }
+                    ResponseEntity<?> responseEntity = (ResponseEntity<?>) payload;
```
Comment (micha1970) [inline application/src/main/java/org/ehrbase/application/abac/CustomMethodSecurityExpressionRoot.java:361]: 358 -364 are the same error message, put them together

File: application/src/main/java/org/ehrbase/application/abac/CustomMethodSecurityExpressionRoot.java [hunk 5]
```diff
@@ -497,14 +545,12 @@ private boolean sendRequestForEach(
...
      */
-    private JwtAuthenticationToken getJwtAuthenticationToken(Authentication auth) {
-        JwtAuthenticationToken jwt;
-        if (auth instanceof JwtAuthenticationToken) {
-            jwt = (JwtAuthenticationToken) auth;
+    private AbstractAuthenticationToken getJwtAuthenticationToken(Authentication auth) {
+        if (auth instanceof AbstractAuthenticationToken jwt) {
+            return jwt;
         } else {
             throw new IllegalArgumentException("ABAC: Invalid authentication, no JWT available.");
```
Comment (micha1970) [inline application/src/main/java/org/ehrbase/application/abac/CustomMethodSecurityExpressionRoot.java:552]: Just remove the esle block and throw the exception, but that's really up to you

File: application/src/main/java/org/ehrbase/application/abac/CustomMethodSecurityExpressionRoot.java (hunk 3 above)
Comment (dmitrytulba) [reply application/src/main/java/org/ehrbase/application/abac/CustomMethodSecurityExpressionRoot.java:260]: claims is different Map<String, Object> and Map<String, Claim>

File: application/src/main/java/org/ehrbase/application/abac/CustomMethodSecurityExpressionRoot.java (hunk 4 above)
Comment (dmitrytulba) [reply application/src/main/java/org/ehrbase/application/abac/CustomMethodSecurityExpressionRoot.java:361]: I changed the message in other pr

File: application/src/main/java/org/ehrbase/application/abac/CustomMethodSecurityExpressionRoot.java (hunk 5 above)
Comment (dmitrytulba) [reply application/src/main/java/org/ehrbase/application/abac/CustomMethodSecurityExpressionRoot.java:552]: I would rather leave it for now. Code will be changed

**Pull request merged**